### PR TITLE
[Windows] Add full support for HDR toggle in Windows 11 24H2

### PR DESCRIPTION
**Pull request by thexai**
## Description
Add full support for HDR toggle in Windows 11 24H2

Make use of new APIs:
- DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO_2
- DISPLAYCONFIG_DEVICE_INFO_SET_HDR_STATE

## Motivation and context
HDR toggle is partially broken in Windows 11 24H2, depending on user Windows settings:

i.e. not works if user enables "Automatically manage color for apps" Windows setting but in my case works same as before in Intel NUC after upgrade to 24H2 and not touch any Windows settings (i.e. current code still works with Windows 11 24H2 default settings).

In forums some users complains HDR toggle not works at all:
https://forum.kodi.tv/showthread.php?tid=379728
https://forum.kodi.tv/showthread.php?tid=379533

New API seems more appropriate for HDR toggle as differentiates "advanced color" or WCG and HDR and implements specific flag to toggle HDR only via `DISPLAYCONFIG_SET_HDR_STATE`.

This only is supported from 24H2, so for older Windows versions is continued used same code as before.

## How has this been tested?
Runtime tested in Windows 11 24H2 --> Reproduced the issue: HDR toggle not works after enable "Automatically manage color for apps" Windows setting.

**Before**
"Automatically manage color for apps" OFF --> HDR works
"Automatically manage color for apps" ON --> HDR not works

**After (this PR)**
"Automatically manage color for apps" OFF --> HDR works
"Automatically manage color for apps" ON --> HDR works

## What is the effect on users?
- Fixes HDR toggle broken in Windows 11 24H2 under some circumstances.
- Fixes issues related to WCG (wide color gamut) screens (only fixed in Windows 11 24H2).

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by thexai** [comment]
**NOTE:**

This PR increases build requirements to Windows SDK `10.0.26100.0` but not has any effect on runtime requirements i.e Kodi continues working in Windows 10.

Then the required SDK should be installed in Jenkins builders:


SDK

**Comment by fuzzard** [comment]
Docs need to be updated to match that new requirement i imagine?

**Comment by CrystalP** [comment]
Confirmed that the new values don't work on Windows 10 so a new path is needed.

Reporting of HDR status by GetWindowsHDRStatus is using different API (PR#24689), maybe it outputs the wrong values too?
If you didn't have trouble with your systems we'll need logs from users.
It's not a good thing to have differences in the Get() and the Toggle() paths.

Maybe the things added in #24689 and UWP could be rolled back if they're useful only for Win 11 22H2. We could use the Win32  / DisplayConfigGetDeviceInfo() way only. In the end we didn't get firm information about WCG mode and screens, which the PR was supposed to help with.

**Comment by CrystalP** [comment]
Sometime in the last year, a change caused debug exe to fail to start on Windows 8.1 which is a pain. Release exe are still ok for some reason (not tested with this PR yet). Runtime and sdk bumps are prime suspects. I've been meaning to track down the commit, this PR adds a bit of motivation.

File: cmake/scripts/windows/ArchSetup.cmake [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Minimum SDK version required to build
-set(VS_MINIMUM_BUILD_SDK_VERSION 10.0.22621.0)
+set(VS_MINIMUM_BUILD_SDK_VERSION 10.0.26100.0)
```
Comment (CrystalP) [inline cmake/scripts/windows/ArchSetup.cmake:2]: What about UWP, bump as well?

Comment (thexai) [reply cmake/scripts/windows/ArchSetup.cmake:2]: Changes in this PR only affects Windows desktop, so there is no reason to require increaser SDK for UWP/Xbox

**Comment by thexai** [comment]
> Sometime in the last year, a change caused debug exe to fail to start on Windows 8.1 which is a pain. Release exe are still ok for some reason (not tested with this PR yet). Runtime and sdk bumps are prime suspects. I've been meaning to track down the commit, this PR adds a bit of motivation.

This comment is not very related to this PR. In any case, for end users, it is much more important that HDR works well in the latest  Windows 11 version than that there are problems, in debug mode only, in Windows 8.1 (a system that is in the process of being discontinued).

**Comment by thexai** [comment]
> Reporting of HDR status by GetWindowsHDRStatus is using different API (PR#24689), maybe it outputs the wrong values too?

Nothing is wrong, I tested this PR on Windows 11 24H2. Tested means tested code changes in this PR and code alredy present in master branch (including  PR#24689).

> Maybe the things added in https://github.com/xbmc/xbmc/pull/24689 and UWP could be rolled back if they're useful only for Win 11 22H2. We could use the Win32 / DisplayConfigGetDeviceInfo() way only. In the end we didn't get firm information about WCG mode and screens, which the PR was supposed to help with.

No, https://github.com/xbmc/xbmc/pull/24689 is fine and although related is a different thing: `GetWindowsHDRStatus` is different of `ToggleWindowsHDR` (one is "get status" the other is "set status"). Same WinRT code path working in Windows 11 desktop and Xbox is very good thing.

Also user issues reported in forums are unrelated because users has v21.1 only and mentioned WinRT PR only is merged in master (not backported).

**Comment by thexai** [comment]
> Sometime in the last year, a change caused debug exe to fail to start on Windows 8.1 which is a pain.

Unrelated: open a Bug issue providing more details.

**Comment by CrystalP** [comment]
> > Sometime in the last year, a change caused debug exe to fail to start on Windows 8.1 which is a pain. Release exe are still ok for some reason (not tested with this PR yet). Runtime and sdk bumps are prime suspects. I've been meaning to track down the commit, this PR adds a bit of motivation.
> 
> This comment is not very related to this PR. In any case, for end users, it is much more important that HDR works well in the latest Windows 11 version than that there are problems, in debug mode only, in Windows 8.1 (a system that is in the process of being discontinued).

Comment prompted by the sdk bump in the PR and the complications that happened with the last bump. Care and investigation is needed to avoid breaking things.

I also think that sdk bumps and Visual Studio 2019 deprecation should not be done without mention in a PR, a dedicated commit would be good for clarity.

**Comment by CrystalP** [comment]
> > Maybe the things added in #24689 and UWP could be rolled back if they're useful only for Win 11 22H2. We could use the Win32 / DisplayConfigGetDeviceInfo() way only.
> 
> No, #24689 is fine and although related is a different thing: `GetWindowsHDRStatus` is different of `ToggleWindowsHDR` (one is "get status" the other is "set status"). Same WinRT code path working in Windows 11 desktop and Xbox is very good thing.

I don't agree, Toggle internally does a Get operation to decide if it should switch on or switch off. There is no reason for GetWindowsHDRStatus and the Get inside Toggle to not match.
They matched until 24689, I didn't notice the issue at the time and have been meaning to refactor. There is also already too much duplication of code and ToggleWindowsHDR_24H2 adds one more copy of the loop to find the display device.

> Also user issues reported in forums are unrelated because users has v21.1 only and mentioned WinRT PR is merged in master only (not backported).

All in all that makes me think that the solutions for v21 and v22 should be different. For v21 the approach of this PR is OK as long as GetWindowsHDRStatus is modified to use the new DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO_2 with Windows 11 24H2 as well. Or make ToggleWindowsHDR_24H2 use the same DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO as GetWindowsHDRStatus .

**Comment by CrystalP** [comment]
> Sometime in the last year, a change caused debug exe to fail to start on Windows 8.1 which is a pain.

replying to myself, this was caused by a debug runtime that needed upgrading on the Windows 8.1 computer. VS doesn't do it with remote debug.

The SDK 1.0.26100.0 bump is fine with Windows 8.1.

on the build machine only the SDK should be bumped and not Visual Studio, as it would likely update the required runtime and cause problems like happened with the last bump.

**Comment by thexai** [comment]
> on the build bot only the SDK should be bumped and not Visual Studio, as it would likely update the required runtime and cause problems like happened with the last bump.

This not makes any sense: with Windows 11 at least Visual Studio is updated in background together Windows updates then is inevitable use always the latest version. This is not causing me any problems.

**Comment by thexai** [comment]
> I don't agree, Toggle internally does a Get operation to decide if it should switch on or switch off. There is no reason for GetWindowsHDRStatus and the Get inside Toggle to not match.
> They matched until 24689, I didn't notice the issue at the time and have been meaning to refactor. There is also already too much duplication of code and ToggleWindowsHDR_24H2 adds one more copy of the loop to find the display device.

This is also _unrelated_ to this PR: different code type happened in 24689 (WinRT for Get and Win32 for Set) then this "issue" is already present in current master.

This PR only replaces "Win32 style Set" with other "Win32 style Set" only for 24H2. The WinRT code was already there before.

> adds one more copy of the loop to find the display device

This PR simplifies loops with continue; and less braces. I plan to make many more simplifications but this PR only includes the minimal changes so that it is very clear what has changed due 24H2.

**Comment by thexai** [comment]
> All in all that makes me think that the solutions for v21 and v22 should be different. For v21 the approach of this PR is OK as long as GetWindowsHDRStatus is modified to use the new DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO_2 with Windows 11 24H2 as well.

That's not going to happen. No changes should be introduced in v21 that are not first tested on master and then backported (same code).

> For v21 the approach of this PR is OK

This PR is for v22 like all others new code PRs

**Comment by thexai** [comment]
OK modified the PR to NOT require Windows SDK bump 

New values defined locally conditional depending if SDK exist or not at time of build.

**Comment by CrystalP** [comment]
OK, have yet to test the sdk declaration but that should work until we figure out how to update the build bot in a controlled manner.

I'm OK to apply a minimal change to v22 that will be backported to v21 and then improve v22 (which will really diverge from v21)

For a minimal change, I haven't understood clearly yet which api fails. No need to change what works.

DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO or DISPLAYCONFIG_DEVICE_INFO_SET_ADVANCED_COLOR_STATE or both?

**Comment by thexai** [comment]
> For a minimal change, I haven't understood clearly yet which api fails. No need to change what works.
> 
> DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO or DISPLAYCONFIG_DEVICE_INFO_SET_ADVANCED_COLOR_STATE or both?

Both:

`DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO` => `DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO_2`

`DISPLAYCONFIG_DEVICE_INFO_SET_ADVANCED_COLOR_STATE` => `DISPLAYCONFIG_DEVICE_INFO_SET_HDR_STATE`

**Comment by CrystalP** [comment]
> For a minimal change, I haven't understood clearly yet which api fails. No need to change what works.
> 
> DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO or DISPLAYCONFIG_DEVICE_INFO_SET_ADVANCED_COLOR_STATE or both?

Which one is broken by 24H2? I don't have access to 24H2 for testing.

>For backport (as always) same code that master:
>
>24689 + https://github.com/xbmc/xbmc/pull/25670 + this PR

24689 was supposed to be v22 only and not backported because it's a significant and risky change.
Because v21 and v22 are different I don't think the usual logic for backports can be used. Let's ask in Slack, it can't be the first time this happens.

**Comment by thexai** [comment]
> 24689 was supposed to be v22 only and not backported because it's a significant and risky change.

But is also a fix (WCG screens issue) and now is tested in master.

Is more risky implement a new code path for 21 only... and release without test it.

**Comment by thexai** [comment]
> Which one is broken by 24H2? I don't have access to 24H2 for testing.

**Both**

DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO => DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO_2

DISPLAYCONFIG_DEVICE_INFO_SET_ADVANCED_COLOR_STATE => DISPLAYCONFIG_DEVICE_INFO_SET_HDR_STATE

**Comment by CrystalP** [comment]
> > Which one is broken by 24H2? I don't have access to 24H2 for testing.
> 
> **Both**
> 
> `DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO => DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO_2`
> 
> `DISPLAYCONFIG_DEVICE_INFO_SET_ADVANCED_COLOR_STATE => DISPLAYCONFIG_DEVICE_INFO_SET_HDR_STATE`

OK, it means that DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO is broken in both Toggle and v21 GetWindowsHDRStatus().
What about the winrt path for GetWindowsHDRStatus in v22, still works or broken as well?

**Comment by thexai** [comment]
> What about the winrt path for GetWindowsHDRStatus in v22, still works or broken as well?

Still works because already differentiate WCG and HDR (not only generic "advanced color").

The new 24H2 API is the equivalent to existent WinRT API for Win32.

**Comment by CrystalP** [comment]
As mentioned before, I'm concerned with consistency. Since the winrt api still works then there are two options to fix 24H2: consistent use of winrt api or consistent use of new win32 api (of course uwp would continue with winrt api, the only one available).
Which one is your preference?

**Comment by thexai** [comment]
My preference is use WinRT for "Get status" because covers WCG issue in Windows 11 22H2 and 23H2 but is not relevant of this PR.

**Comment by thexai** [comment]
Added HDR status detection based in new Win32 API only for Windows 11 24H2.

WinRT detection path is still necessary for protect Win11 22H2 and 23H2 from WCG issue.

I'm fine to do "special" backport of only new 24H2 code (not include WinRT HDR detection for desktop) if you considers is still experimental or not stable.

**Comment by CrystalP** [comment]
Since this wasn't going anywhere I wrote the PR #26129 to bring the code of Get and Toggle back in sync and hopefully things are clearer now.
The Get function is fine as is since you confirmed that it still works with 24H2 and ACG > no change needed for this PR.
Toggle only needs to add the DISPLAYCONFIG_DEVICE_INFO_SET_HDR_STATE win32 api in the new SetWindowsHDRStatus (or a 24H2 version of it)
Since winrt is your preference and I don't care that much one way or the other.

**Comment by thexai** [comment]
Follow-up PRs come after this one. We have to agree on this one first and when it's merged you make whatever improvements you want.  

That seem fair to me.

**Comment by thexai** [comment]
> Maybe the things added in https://github.com/xbmc/xbmc/pull/24689 and UWP could be rolled back if they're useful only for Win 11 22H2. We could use the Win32 / DisplayConfigGetDeviceInfo() way only. In the end we didn't get firm information about WCG mode and screens, which the PR was supposed to help with.

I changed my mind, this is also my preference currently. Revert https://github.com/xbmc/xbmc/pull/24689 and simplify things... so use Win32 API only on desktop (with 24H2 compatibility) and backport 1:1 to Omegle

**Comment by CrystalP** [comment]
My eyes hurt from all the code duplication, but that's the fix with the lowest risk when thinking about a backport, and going forward as a base for refactor and prettyification.

Is the purpose of DISPLAYCONFIG_GET_ADVANCED_COLOR_INFO_2/wideColorSupported understood?
What about DISPLAYCONFIG_SET_WCG_STATE? Maybe they both map to the Auto Color Management setting?

**Comment by thexai** [comment]
> My eyes hurt from all the code duplication, but that's the fix with the lowest risk when thinking about a backport, and going forward as a base for refactor and prettyification.

Git diff looks bad :( but changes are simple if you look full code.

> Is the purpose of DISPLAYCONFIG_GET_ADVANCED_COLOR_INFO_2/wideColorSupported understood?
> What about DISPLAYCONFIG_SET_WCG_STATE? Maybe they both map to the Auto Color Management setting?

`DISPLAYCONFIG_SET_WCG_STATE` is for set/toggle WCG (similar to toggle HDR) but not used in Kodi code.

`wideColorSupported` understood? --> Display supports WCG (regardless of whether it supports HDR).

**Comment by CrystalP** [comment]
it's cheating, the revert is included in the line count :-)
anyway I prepared something to refactor, will rebase on top of this.

**Review by CrystalP — APPROVED** [review]
Best course of action at this point, though more work will be needed to put this code in good shape.

**Comment by thexai** [comment]
For me is almost the end of day --> merge it tomorrow after do some tests (I want test in Xbox to make sure no side effects, especially for backport).

**Comment by CrystalP** [comment]
Things got more "interesting". I don't use Win11 on a regular basis and didn't notice earlier, but with 23H2 I found exactly the same symptoms as the user who prompted the winrt code addition, for the laptop's internal and external screens (Kodi thinks they're all HDR capable, though they're not), and turning ACM on does make Kodi think that HDR is on (it's not). I also find that the same win32 fix imagined back then, using wideColorEnforced of DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO works as well, which makes me more confident in that solution.
So while using the new APIs for 24H2 is still better and this PR doesn't have to change, I'll submit a fix for 22H2 and 23H2 and the incorrect HDR detection and there won't be any gaps in the Windows versions supported.

**Pull request merged**